### ISSUE TITLE
Troubles instantiating xmlschema.XMLSchema class (ReqIF XML format)

### ISSUE THREAD

**Issue by stanislaw** [sissaschool/xmlschema#360]
Hello,

TL;DR: Instantiating `xmlschema.XMLSchema` hangs somewhere in the Python's library socket code, e.g., `sock.connect(sa)`. I apologize if I am misusing the API and will be grateful for any advice.

```py
import xmlschema
schema = xmlschema.XMLSchema(
     "https://www.omg.org/spec/ReqIF/20110401/reqif.xsd"
)
```

The detailed report is below.

---

Thank you for creating this library. I am a developer of the ReqIF library called [reqif](https://github.com/strictdoc-project/reqif) Python library, which is a child project of my [StrictDoc](https://github.com/strictdoc-project/strictdoc/pulls) requirements management tool.

The ReqIF library has reached a good state of maturity, and I would like to extend it with a capability to validate the ReqIF XML files using the official XSD file provided [here](https://www.omg.org/spec/ReqIF/20110401/reqif.xsd).

I am trying to do something as simple as follows:

```xml
import xmlschema
schema = xmlschema.XMLSchema(
    "https://www.omg.org/spec/ReqIF/20110401/reqif.xsd"
)
```

I also tried to download a local copy of this file, but the result is the same: in both cases, the script hangs with no output or any indication of progress, the program does not terminate, unless I Ctrl-C it.

```
schema = xmlschema.XMLSchema(
    "reqif.xsd"
)
```

I reduced the example to the minimum reproducing one, and that is:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsd:schema xmlns="http://www.w3.org/1999/xhtml" xmlns:REQIF="http://www.omg.org/spec/ReqIF/20110401/reqif.xsd" xmlns:xml="http://www.w3.org/XML/1998/namespace" xmlns:xsd="http://www.w3.org/2001/XMLSchema" attributeFormDefault="unqualified" elementFormDefault="qualified" targetNamespace="http://www.omg.org/spec/ReqIF/20110401/reqif.xsd">
  <xsd:import
    namespace="http://www.w3.org/1999/xhtml"
    schemaLocation="http://www.omg.org/spec/ReqIF/20110402/driver.xsd"
  />
</xsd:schema>
```

The part that causes the `xmlschema` to hang is this: **schemaLocation="http://www.omg.org/spec/ReqIF/20110402/driver.xsd"**.

When I terminate the program, the stack trace I get is this:

```
^CTraceback (most recent call last):
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/Stanislaw/workspace/projects/strictdoc-project/reqif/reqif/cli/main.py", line 52, in <module>
    main()
  File "/Users/Stanislaw/workspace/projects/strictdoc-project/reqif/reqif/cli/main.py", line 44, in main
    ValidateCommand.execute(parser.get_validate_config())
  File "/Users/Stanislaw/workspace/projects/strictdoc-project/reqif/reqif/commands/validate/validate.py", line 44, in execute
    error_bundle = ValidateCommand._validate(config)
  File "/Users/Stanislaw/workspace/projects/strictdoc-project/reqif/reqif/commands/validate/validate.py", line 65, in _validate
    "https://www.omg.org/spec/ReqIF/20110401/reqif.xsd"
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/xmlschema/validators/schemas.py", line 346, in __init__
    self.source = XMLResource(source, base_url, allow, defuse, timeout)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/xmlschema/resources.py", line 471, in __init__
    self.parse(source, lazy)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/xmlschema/resources.py", line 732, in parse
    with urlopen(url, timeout=self._timeout) as resource:
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 543, in _open
    '_open', req)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1393, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1350, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1281, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1327, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1276, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1036, in _send_output
    self.send(msg)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 976, in send
    self.connect()
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 1443, in connect
    super().connect()
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/http/client.py", line 948, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/Users/Stanislaw/.pyenv/versions/3.7.15/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/socket.py", line 716, in create_connection
    sock.connect(sa)
KeyboardInterrupt
```

Besides the fact that `xmlschema` is hanging on an URL that is known to exist, e.g., http://www.omg.org/spec/ReqIF/20110402/driver.xsd, I am wondering why does `xmlschema` have to visit the internet in order to proceed with the validation? I would like to be able to keep all the necessary XSD files bundled with my tool and be able to have a fast off-line validation of the user files according to the given XSD schema. Is this possible with `xmlschema` to bundle all XSDs locally?

I hope that my report is clear, and I am willing to provide further information.

Thank you.

Stanislav

**Comment by brunato** [comment]
Hi,

the loading of schema works for me (Python 3.7.17):

```pycon
>>> import xmlschema
>>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
>>> schema.imports
{'http://www.w3.org/XML/1998/namespace': MetaXMLSchema10(name='xml_minimal.xsd', namespace='http://www.w3.org/XML/1998/namespace'), 'http://www.w3.org/1999/xhtml': XMLSchema10(name='driver.xsd', namespace='http://www.w3.org/1999/xhtml')}
```

I think that your error doesn't depend by _xmlschema_ nor Python's _urllib_ but it should be a connection problem (temporary or not). Note that the xml namespace ('http://www.w3.org/XML/1998/namespace') is loaded from a local fallback schema, because the location _https://www.w3.org/2001/xml.xsd_ does not contain a valid XSD schema.

Obviously it's possible to bundle a set of local loading schemas downloading them from remote and then changing the schemaLocation attributes into _xs:import/xs:include_ statements.

If remote schemas are addressed only for imports (that is the best practice that xs:include statements use only local URLs) you can download remote schema sets and then provide the argument _locations_ to `XMLSchema()` to override _schemaLocation_ attribute in `xs:import` statements.

Also, having a loaded schema instance, you can export the schema to a local directory, e.g.:

```pycon
>>> schema.export(target='my_schemas', save_remote=True)
```

Best regards

Davide

**Comment by stanislaw** [comment]
> Hi,

Thanks for a quick answer!

> the loading of schema works for me (Python 3.7.17):
> 
> ```
> >>> import xmlschema
> >>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
> >>> schema.imports
> {'http://www.w3.org/XML/1998/namespace': MetaXMLSchema10(name='xml_minimal.xsd', namespace='http://www.w3.org/XML/1998/namespace'), 'http://www.w3.org/1999/xhtml': XMLSchema10(name='driver.xsd', namespace='http://www.w3.org/1999/xhtml')}
> ```
> 
> I think that your error doesn't depend by _xmlschema_ nor Python's _urllib_ but it should be a connection problem (temporary or not). 

I tried a pure Python console approach, and I am getting this - it definitely hangs:

```py
>>> import xmlschema
>>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
```

I went for a 5-min coffee break, and it eventually worked but after at least more than 60 seconds. My internet connection is perfect, and I am able to download all these files using the browser in milliseconds. I ran it again and it is hanging while I am writing this whole answer, definitely more than 5 minutes so far.

I don't know why it takes 60+ seconds to get a file, but actually I would prefer to not even look into this because a local workflow would be much more optimal. My questions about the local workflow are below.

> Note that the xml namespace ('http://www.w3.org/XML/1998/namespace') is loaded from a local fallback schema, because the location _https://www.w3.org/2001/xml.xsd_ does not contain a valid XSD schema.
> 
> Obviously it's possible to bundle a set of local loading schemas downloading them from remote and then changing the schemaLocation attributes into _xs:import/xs:include_ statements.
> 
> If remote schemas are addressed only for imports (that is the best practice that xs:include statements use only local URLs) you can download remote schema sets and then provide the argument _locations_ to `XMLSchema()` to override _schemaLocation_ attribute in `xs:import` statements.
> 
> Also, having a loaded schema instance, you can export the schema to a local directory, e.g.:
> 
> ```
> >>> schema.export(target='my_schemas', save_remote=True)
> ```

I am new to the XSD language, so I cannot follow your explanation 100% just yet. Could you help me understand this with a simple "yes" or "no" answer for the immediate time being: is it possible to bundle all the xsd file's dependencies locally and be able **without modifying any of them** make `xmlschema` to validate my XML file 100% locally without going to the internet?

My use case is very clear: if I enable validation of ReqIF files using `xmlschema`, I want to be able to run the validation fast and deterministic because the users and I will most likely need to run it on CI and also fast but locally. This is why the 100% local workflow is a MUST requirement for the whole XSD validation feature to work in my case.

> >>> schema.export(target='my_schemas', save_remote=True)

This rings a bell with respect to a local workflow I am looking for, but which command should I use to read this exported information without going to the internet?

Thank you.

**Comment by brunato** [comment]
> I tried a pure Python console approach, and I am getting this - it definitely hangs:
> 
> ```python
> >>> import xmlschema
> >>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
> ```
> 
> I went for a 5-min coffee break, and it eventually worked but after at least more than 60 seconds. My internet connection is perfect, and I am able to download all these files using the browser in milliseconds. I ran it again and it is hanging while I am writing this whole answer, definitely more than 5 minutes so far.
> 
> I don't know why it takes 60+ seconds to get a file, but actually I would prefer to not even look into this because a local workflow would be much more optimal. My questions about the local workflow are below.

_xmlschema_ simply uses the Python `urllib.request.urlopen` call, so you can try if this works fine instead:

```pycon
>>> from urllib.request import urlopen
>>> response = urlopen("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
>>> response.read()
```

> I am new to the XSD language, so I cannot follow your explanation 100% just yet. Could you help me understand this with a simple "yes" or "no" answer for the immediate time being: is it possible to bundle all the xsd file's dependencies locally and be able **without modifying any of them** make `xmlschema` to validate my XML file 100% locally without going to the internet?

No, you have to modify the `xs:import` and `xs:include` statements of original XSD schemas.


> > > > > schema.export(target='my_schemas', save_remote=True)
> 
> This rings a bell with respect to a local workflow I am looking for, but which command should I use to read this exported information without going to the internet?

If you have for example:

```python
>>> import xmlschema
>>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
>>> schema.export(target='my_schemas', save_remote=True)
```

then you can build the same schema instance from local files with:

```pycon
>>> schema = xmlschema.XMLSchema("my_schemas/reqif.xsd")
```

**Comment by stanislaw** [comment]
> > I tried a pure Python console approach, and I am getting this - it definitely hangs:
> > ```python
> > >>> import xmlschema
> > >>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
> > ```
> > I went for a 5-min coffee break, and it eventually worked but after at least more than 60 seconds. My internet connection is perfect, and I am able to download all these files using the browser in milliseconds. I ran it again and it is hanging while I am writing this whole answer, definitely more than 5 minutes so far.
> > I don't know why it takes 60+ seconds to get a file, but actually I would prefer to not even look into this because a local workflow would be much more optimal. My questions about the local workflow are below.
> 
> _xmlschema_ simply uses the Python `urllib.request.urlopen` call, so you can try if this works fine instead:
> 
> ```
> >>> from urllib.request import urlopen
> >>> response = urlopen("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
> >>> response.read()
> ```

I tried Python 3.7 and Python 3.11. I am on `macOS 12.6.3 (21G419)`. I confirm that the urlopen lines above work for me right away, and the response is read within a fraction of a second. Doing the same with `xmlschema.XMLSchema` hangs for 3–5 minutes like in my original report. I managed to set the local workflow like you suggested, see below, so I can give up on investigating this further, unless you would want me to try any debugging tricks or other users would report something similar.

> > I am new to the XSD language, so I cannot follow your explanation 100% just yet. Could you help me understand this with a simple "yes" or "no" answer for the immediate time being: is it possible to bundle all the xsd file's dependencies locally and be able **without modifying any of them** make `xmlschema` to validate my XML file 100% locally without going to the internet?
> 
> No, you have to modify the `xs:import` and `xs:include` statements of original XSD schemas.
> 
> > > > > > schema.export(target='my_schemas', save_remote=True)
> > 
> > 
> > This rings a bell with respect to a local workflow I am looking for, but which command should I use to read this exported information without going to the internet?
> 
> If you have for example:
> 
> ```python
> >>> import xmlschema
> >>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
> >>> schema.export(target='my_schemas', save_remote=True)
> ```
> 
> then you can build the same schema instance from local files with:
> 
> ```
> >>> schema = xmlschema.XMLSchema("my_schemas/reqif.xsd")
> ```

I did manage to run the validation using the local workflow based on your instructions. Now that the local feature works, I tried to find these instructions in the existing documentation and could not find it. I am happy to send a patch that would add a section "How to enable an offline validation of XSD files" and basically but there would you explained to me. Do you think you would benefit from such a patch? Let me know.

```
import xmlschema
schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd")
schema.export(target='my_schemas', save_remote=True)
schema = xmlschema.XMLSchema("my_schemas/reqif.xsd")  # works without internet 👍 
```

**Comment by stanislaw** [comment]
One issue I noticed was that one file `xml_minimal.xsd` was generated under a full path of my user's Python dirs. I deleted this file in my setup and noticed that it didn't break my workflow, so I am not affected by this issue but maybe you would want to fix it for your general tool application.

<img width="463" alt="Bildschirmfoto 2023-07-31 um 21 46 42" src="https://github.com/sissaschool/xmlschema/assets/452547/06f32348-c381-45b7-835a-56a927a38e05">

**Comment by brunato** [comment]
> I tried Python 3.7 and Python 3.11. I am on `macOS 12.6.3 (21G419)`. I confirm that the urlopen lines above work for me right away, and the response is read within a fraction of a second. Doing the same with `xmlschema.XMLSchema` hangs for 3–5 minutes like in my original report. I managed to set the local workflow like you suggested, see below, so I can give up on investigating this further, unless you would want me to try any debugging tricks or other users would report something similar.

You can also try debugging the schema instance build:

```pycon
>>> import xmlschema
>>> import logging
>>> logger = logging.getLogger('xmlschema') 
>>> logger.addHandler(logging.StreamHandler(sys.stderr))
>>> schema = xmlschema.XMLSchema("https://www.omg.org/spec/ReqIF/20110401/reqif.xsd", loglevel=logging.DEBUG)
```

> I did manage to run the validation using the local workflow based on your instructions. Now that the local feature works, I tried to find these instructions in the existing documentation and could not find it. I am happy to send a patch that would add a section "How to enable an offline validation of XSD files" and basically include there what you explained to me. Do you think you would benefit from such a patch? Let me know.

Happy if you can do this for adding a section for export and offline validation into documentation.

Thank you

**Comment by stanislaw** [comment]
By now, I have integrated `xmlschema` into my tool, and it is working perfectly 👌 

I will get to opening a documentation PR during the next weekend. I will also see if I can find out why this huge delay occurs when I validate the remote xsd. I will come back to this issue with the further feedback. 

Thank you for your quick answers.

**Comment by stanislaw** [comment]
I am closing this issue because I found out that some other packages hang when they open network connections. The problem is clearly on my side, I just could not figure out yet what is the cause.

The most important outcome of this issue though for me was the offline workflow documented in #361. 

Thank you for your attention to this issue.
